crates/pessimistic-proof-core/src/aggchain_proof.rs

```diff
@@ -0,0 +1,103 @@
+//! Aggchain proof data structures.
+//!
+//! The aggchain-proof is the extra custom logic which is verified within the
+//! pessimistic-proof.
+//!
+//! For now, this is constraint to be either one ECDSA signature, or one SP1
+//! stark proof proving a specified statement which can be abstracted here.
+use agglayer_primitives::{Address, Signature};
+use serde::{Deserialize, Serialize};
+use sha2::{Digest as Sha256Digest, Sha256};
+use sp1_zkvm::lib::utils::words_to_bytes_le;
+
+use crate::keccak::{digest::Digest, keccak256_combine};
+
+#[derive(Debug, Deserialize, Serialize, Clone, Copy)]
+pub enum AggchainType {
+    ECDSA = 0,
+    SP1 = 1,
+}
+
+pub type Vkey = [u32; 8];
+pub type StarkVkey = Vec<u8>;
+pub type StarkProof = Vec<u8>;
+
+/// Aggchain Proof which is either one ECDSA signature, or one stark proof.
+/// Contains all the necessary data for verification.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum AggchainProofData {
```

Comment:
Shouldn't it be just `AggchainProof` here?

Reply:
I wanted to have the `AggchainProof` name for the structure in the `Certificate` which includes only 
- if SP1 proof: `proof` + `aggchain_params`
- if ECDSA: `signature`


While `AggchainProofData` would be the entire aggchain proof information given to the prover, so:

- if SP1 proof: includes also the `vkey` retrieved from L1
- if ECDSA: includes also the `signer` retrieved from L1

Reply:
I would have preferred to have `AggchainProof` and `AggchainProofVerifier` that clearly defined what is part of the `aggchain-proof` and what is needed to verify it.
At some point if we're moving the `pessimistic-proof-core` and `program` to another repo, we'll maybe loose this context.

Another solution could be to have something like:

```rs
enum AggchainProof {
	ECDSA(..),
	SP1(..)
}

enum AggchainProofVerifier {
	ECDSA(..),
	SP1(..)
}

enum AggchainProofWithVerifier {
	proof: AggchainProof,
	verifier: AggchainProofVerifier
}
```

Reply:
> Another solution could be to have something like:
> 
> ```rust
> enum AggchainProof {
> 	ECDSA(..),
> 	SP1(..)
> }
> 
> enum AggchainProofVerifier {
> 	ECDSA(..),
> 	SP1(..)
> }
> 
> enum AggchainProofWithVerifier {
> 	proof: AggchainProof,
> 	verifier: AggchainProofVerifier
> }
> ```

Ideally, the types make nonsensical states impossible. The `AggchainProofWithVerifier` suggested above makes it possible to mix together ECDSA "challenge" and SP1 "witness data" and vice versa, whereas the current representation in the PR does not.